c++/src/capnp/serialize-async.c++

```diff
@@ -502,4 +504,288 @@ kj::Promise<MessageReaderAndFds> MessageStream::readMessage(
+}
+
+kj::Maybe<BufferedAsyncIoMessageStream::IncompleteRead>
+    BufferedAsyncIoMessageStream::tryGetIncompleteRead() {
+  return kj::mv(incompleteRead);
+}
+
+void BufferedAsyncIoMessageStream::bufferCompleteMessage(MessageReaderAndOwnedFds message) {
+  bufferedReaders.push_back(kj::mv(message.reader));
+}
+
+void BufferedAsyncIoMessageStream::bufferIncompleteMessage(
+    kj::ArrayPtr<kj::byte> leftover, kj::Array<kj::AutoCloseFd> fds, size_t size) {
+  kj::ArrayPtr<word> leftoverWords(reinterpret_cast<word*>(leftover.begin()),
+      leftover.size() / sizeof(word));
+
+  auto& incomplete = incompleteRead.emplace(IncompleteRead {
+    kj::heapArray<byte>(leftover.size()),
+    capnp::expectedSizeInWordsFromPrefix(leftoverWords) * sizeof(word)
+  });
+  memcpy(incomplete.bytes.begin(), leftover.begin(), leftover.size());
+}
+
+kj::Promise<BufferedAsyncCapabilityMessageStream::ReadResult>
+    BufferedAsyncCapabilityMessageStream::tryRead(kj::Array<byte> buffer, void* start,
+      size_t minBytes, size_t maxBytes) {
+  auto fdSpace = kj::heapArray<kj::AutoCloseFd>(options.maxFdsPerMessage);
+  return stream.tryReadWithFds(start, minBytes, maxBytes, fdSpace.begin(),
+      options.maxFdsPerMessage).then(
+      [this, buffer = kj::mv(buffer), fdSpace = kj::mv(fdSpace)]
+      (kj::AsyncCapabilityStream::ReadResult result) mutable {
+    return ReadResult(kj::mv(buffer), kj::mv(fdSpace), kj::mv(incompleteReadFds),
+        result.byteCount, result.capCount);
+  });
+}
+
+kj::Maybe<MessageReaderAndFds> BufferedAsyncCapabilityMessageStream::tryGetBufferedMessage(
+    kj::ArrayPtr<kj::AutoCloseFd> fdSpace) {
+  if (bufferedReaders.size() > 0) {
+    auto next = kj::mv(bufferedReaders.front());
+    bufferedReaders.pop_front();
+
+    for(auto i : kj::zeroTo(next.fds.size())) {
+      if (i >= fdSpace.size()) {
+        break;
+      }
+      fdSpace[i] = kj::mv(next.fds[i]);
+    }
+
+
+    return MessageReaderAndFds {
+      kj::mv(next.reader),
+      fdSpace.slice(0, kj::min(fdSpace.size(), next.fds.size()))
+    };
+  } else {
+    return nullptr;
+  }
+}
+
+kj::Maybe<BufferedAsyncCapabilityMessageStream::IncompleteRead>
+    BufferedAsyncCapabilityMessageStream::tryGetIncompleteRead() {
+  return kj::mv(incompleteRead);
+}
+
+void BufferedAsyncCapabilityMessageStream::bufferCompleteMessage(MessageReaderAndOwnedFds message) {
+  bufferedReaders.push_back(kj::mv(message));
+}
+
+void BufferedAsyncCapabilityMessageStream::bufferIncompleteMessage(
+    kj::ArrayPtr<kj::byte> leftover, kj::Array<kj::AutoCloseFd> fds, size_t size) {
+  kj::ArrayPtr<word> leftoverWords(reinterpret_cast<word*>(leftover.begin()),
+      leftover.size() / sizeof(word));
+
+  auto& incomplete = incompleteRead.emplace(IncompleteRead {
+    kj::heapArray<byte>(leftover.size()),
+    capnp::expectedSizeInWordsFromPrefix(leftoverWords) * sizeof(word)
+  });
+  memcpy(incomplete.bytes.begin(), leftover.begin(), leftover.size());
+  incompleteReadFds = kj::mv(fds);
```

Comment:
Looking at all the private helper methods called by the template function (`tryGetIncompleteRead`, `tryGetBufferedMessage`, `bufferCompleteMessage`, `bufferIncompleteMessage`), it looks like the only actual differences in the implementation between the two classes are this line of code, and the body of `tryGetBufferedMessage`. Is that right? Can we inline more of this code into the shared template function to make it more readable?